presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressions.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.relational;
+
+import com.facebook.presto.metadata.FunctionManager;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression.Form;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.OR;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.google.common.base.Predicates.not;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public final class RowExpressions
+{
+    public static final ConstantExpression TRUE = new ConstantExpression(true, BOOLEAN);
+    public static final ConstantExpression FALSE = new ConstantExpression(false, BOOLEAN);
+
+    private final DeterminismEvaluator determinismEvaluator;
+
+    public RowExpressions(FunctionManager functionManager)
+    {
+        this.determinismEvaluator = new DeterminismEvaluator(requireNonNull(functionManager, "functionManager is null"));
+    }
+
+    public static List<RowExpression> extractConjuncts(RowExpression expression)
+    {
+        return extractPredicates(AND, expression);
+    }
+
+    public static List<RowExpression> extractDisjuncts(RowExpression expression)
+    {
+        return extractPredicates(OR, expression);
+    }
+
+    public static List<RowExpression> extractPredicates(RowExpression expression)
+    {
+        if (expression instanceof SpecialFormExpression) {
```

Comment:
Is this equivalent to
```
List<RowExpression> conjuncts = extractConjuncts(expression);
if (!conjuncts.empty()) {
    return conjuncts;
}
return extractDisjuncts(expression);
```
?

Reply:
Maybe not; if `expression` is a disjunctive, then the entire `expression` will be returned rather than having two disjuncts returned.